### ISSUE TITLE
JInspector

### ISSUE THREAD

**Issue by faustus123** [JeffersonLab/JANA2#112]
Ran into several issues trying to use the built-in JInspector tool from gdb. These were done using the JTest plugin that comes with JANA2. I tried ran it like this:

**gdb --args jana -Pplugins=JTest -Pjana:timeout=600 -Pnthreads=1**
(gdb) **break JTestTracker::Process**
Function "JTestTracker::Process" not defined.
Make breakpoint pending on future shared library load? (y or [n]) **y**
Breakpoint 1 (JTestTracker::Process) pending.
(gdb) **run**

1. If debug symbols are not present, then aEvent is optimized out. The real issue here is whether debug symbols should be included in the default JTest plugin and any plugins created with jana-generate.py. At least for JTest, having them there by default avoids having to document in the tutorial how to re-build JANA with them in the section on running with gdb.
 
2. Even with debug symbols turned on, the JEvent::Inspect() method is inlined so gdb can't call it.

3. Calling _aEvent->mInspector.Loop()_ directly works, but not quite. The initial call to _PrintEvent()_ seems to work before entering the loop and the "JANA:" prompt is printed. Then it seems to revert to printing the ticker line while not responding to any commands. Some temporary print statements seem to indicate it is stalling on the "std::getline(m_in, user_input)" call. Why that frees up the heartbeat thread is unclear. Why it does not accept a command after typing and hitting return is also unclear.

**Comment by nathanwbrei** [comment]
1.  Running gdb on a program without debug symbols is an exercise in frustration. Also, CMake doesn't like it when you override the debug symbols and optimization level for an individual target. There's a CMake flag to compile the whole project as either "Debug", "Release", or "RelWithDebInfo".
  2.  I wonder if there is a way to force the compiler to not inline JEvent::Inspect(). I hope so.
  3.  When using the Inspector I always turned off the heartbeat and the ticker (note these are separate things)

Nathan
________________________________
From: David Lawrence ***@***.***>
Sent: Friday, June 24, 2022 3:23 PM
To: JeffersonLab/JANA2 ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] [JeffersonLab/JANA2] JInspector (Issue #112)


Ran into several issues trying to use the built-in JInspector tool from gdb. These were done using the JTest plugin that comes with JANA2. I tried ran it like this:

gdb --args jana -Pplugins=JTest -Pjana:timeout=600 -Pnthreads=1
(gdb) break JTestTracker::Process
Function "JTestTracker::Process" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (JTestTracker::Process) pending.
(gdb) run

  1.  If debug symbols are not present, then aEvent is optimized out. The real issue here is whether debug symbols should be included in the default JTest plugin and any plugins created with jana-generate.py. At least for JTest, having them there by default avoids having to document in the tutorial how to re-build JANA with them in the section on running with gdb.

  2.  Even with debug symbols turned on, the JEvent::Inspect() method is inlined so gdb can't call it.

  3.  Calling aEvent->mInspector.Loop() directly works, but not quite. The initial call to PrintEvent() seems to work before entering the loop and the "JANA:" prompt is printed. Then it seems to revert to printing the ticker line while not responding to any commands. Some temporary print statements seem to indicate it is stalling on the "std::getline(m_in, user_input)" call. Why that frees up the heartbeat thread is unclear. Why it does not accept a command after typing and hitting return is also unclear.

**Comment by faustus123** [comment]
For 3. I think the problem isn't really the ticker lines (though we should probably disable them while in the inspector loop). This real issue is that it is not accepting commands. What I expect to see is the command and its output interspersed with the ticker lines. I see no indication it ever returns from the getline call.

**Comment by faustus123** [comment]
I take that back. It does seem to be responding to the commands. Somehow fooled myself before.

**Comment by nathanwbrei** [comment]
You might not have been hallucinating. I have a healthy fear of doing I/O from multiple threads concurrently when they themselves call sleep() and there's both an OS scheduler and JANA scheduler to worry about. I know there's a right way to do it but I'm fuzzy on what the right way is. On top of that, TTYs are weird in their own right, and within gdb makes them even weirder. It's possible a thread woke up wrong and somebody didn't give it its tty back.

________________________________
From: David Lawrence ***@***.***>
Sent: Friday, June 24, 2022 3:35 PM
To: JeffersonLab/JANA2 ***@***.***>
Cc: Nathan Brei ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [JeffersonLab/JANA2] JInspector (Issue #112)


I take that back. It does seem to be responding to the commands. Somehow fooled myself before.

**Comment by faustus123** [comment]
OK. I think there was some operator error before. I seem to have it working now. At least if I call aEvent->mInspector.Loop(). I have modified the Loop routine to turn off both the timeout check and ticker before entering the loop and to turn them back on when leaving. It is not ideal since they should really be returned to whatever state they were in when entering the method. JApplication though does not give access to this info at the moment.

**Comment by nathanwbrei** [comment]
That's a good idea, we should add that.
________________________________
From: David Lawrence ***@***.***>
Sent: Friday, June 24, 2022 3:57 PM
To: JeffersonLab/JANA2 ***@***.***>
Cc: Nathan Brei ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [JeffersonLab/JANA2] JInspector (Issue #112)


OK. I think there was some operator error before. I seem to have it working now. At least if I call aEvent->mInspector.Loop(). I have modified the Loop routine to turn off both the timeout check and ticker before entering the loop and to turn them back on when leaving. It is not ideal since they should really be returned to whatever state they were in when entering the method. JApplication though does not give access to this info at the moment.

**Comment by faustus123** [comment]
Yes, if it is fairly easy (e.g. #pragma). If not, then we should make it clearer, at least in embedded documentation in the code, how to use it if it has been optimized out.
